File: day12/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"regexp"
	"strconv"
)

const (
	puzzleInput = "input.txt"
)

type (
	Action struct {
		act string
		v   int
	}

	Ship struct {
		dir string
		x   int
		y   int
		wx  int
		wy  int
	}
)

func NewShip() *Ship {
	return &Ship{
		dir: "E",
		x:   0,
		y:   0,
		wx:  10,
		wy:  -1,
	}
}

func (s *Ship) turnLeft() {
	switch s.dir {
	case "N":
		s.dir = "W"
	case "E":
		s.dir = "N"
	case "S":
		s.dir = "E"
	case "W":
		s.dir = "S"
	}
}

func (s *Ship) turnRight() {
	switch s.dir {
	case "N":
		s.dir = "E"
	case "E":
		s.dir = "S"
	case "S":
		s.dir = "W"
	case "W":
		s.dir = "N"
	}
}

func (s *Ship) Step(a Action) {
	switch a.act {
	case "N":
		s.y -= a.v
	case "S":
		s.y += a.v
	case "E":
		s.x += a.v
	case "W":
		s.x -= a.v
	case "L":
		turns := a.v / 90
		for i := 0; i < turns; i++ {
			s.turnLeft()
		}
	case "R":
		turns := a.v / 90
		for i := 0; i < turns; i++ {
			s.turnRight()
		}
	case "F":
		switch s.dir {
		case "N":
			s.y -= a.v
		case "E":
			s.x += a.v
		case "S":
			s.y += a.v
		case "W":
			s.x -= a.v
		}
	}
}

func (s *Ship) turnWaypointLeft() {
	a := s.wx
	b := s.wy
	s.wx = b
	s.wy = -a
}

func (s *Ship) turnWaypointRight() {
	a := s.wx
	b := s.wy
	s.wx = -b
	s.wy = a
}

func (s *Ship) StepWaypoint(a Action) {
	switch a.act {
	case "N":
		s.wy -= a.v
	case "S":
		s.wy += a.v
	case "E":
		s.wx += a.v
	case "W":
		s.wx -= a.v
	case "L":
		turns := a.v / 90
		for i := 0; i < turns; i++ {
			s.turnWaypointLeft()
		}
	case "R":
		turns := a.v / 90
		for i := 0; i < turns; i++ {
			s.turnWaypointRight()
		}
	case "F":
		s.x += s.wx * a.v
		s.y += s.wy * a.v
	}
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatal(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatal(err)
		}
	}()

	re := regexp.MustCompile(`^([A-Z])([0-9]+)$`)

	s := NewShip()
	s2 := NewShip()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		m := re.FindStringSubmatch(scanner.Text())
		if m == nil {
			log.Fatal("Invalid line")
		}
		act := m[1]
		num, err := strconv.Atoi(m[2])
		if err != nil {
			log.Fatal(err)
		}
		a := Action{
			act: act,
			v:   num,
		}
		s.Step(a)
		s2.StepWaypoint(a)
	}
	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	fmt.Println("Part 1:", abs(s.x)+abs(s.y))
	fmt.Println("Part 2:", abs(s2.x)+abs(s2.y))
}

File: day12/src/main.rs
use regex::Regex;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

type BErr = Box<dyn std::error::Error>;

#[derive(Clone, Copy, PartialEq, Eq)]
enum Action {
    N,
    S,
    E,
    W,
    L,
    R,
    F,
}

struct Instr(Action, i32);

#[derive(Clone, Copy, PartialEq, Eq)]
enum Dir {
    N,
    E,
    S,
    W,
}

impl Dir {
    fn left(&self) -> Dir {
        match self {
            Dir::N => Dir::W,
            Dir::E => Dir::N,
            Dir::S => Dir::E,
            Dir::W => Dir::S,
        }
    }

    fn right(&self) -> Dir {
        match self {
            Dir::N => Dir::E,
            Dir::E => Dir::S,
            Dir::S => Dir::W,
            Dir::W => Dir::N,
        }
    }
}

struct Ship {
    dir: Dir,
    x: i32,
    y: i32,
    wx: i32,
    wy: i32,
}

impl Ship {
    fn new() -> Ship {
        Ship {
            dir: Dir::E,
            x: 0,
            y: 0,
            wx: 10,
            wy: -1,
        }
    }

    fn step(&mut self, a: &Instr) {
        let Instr(act, v) = a;
        match act {
            Action::N => self.y -= v,
            Action::S => self.y += v,
            Action::E => self.x += v,
            Action::W => self.x -= v,
            Action::L => {
                for _ in 0..(v / 90) {
                    self.dir = self.dir.left();
                }
            }
            Action::R => {
                for _ in 0..(v / 90) {
                    self.dir = self.dir.right();
                }
            }
            Action::F => match self.dir {
                Dir::N => self.y -= v,
                Dir::E => self.x += v,
                Dir::S => self.y += v,
                Dir::W => self.x -= v,
            },
        }
    }

    fn turn_waypoint_left(&mut self) {
        let a = self.wx;
        let b = self.wy;
        self.wx = b;
        self.wy = -a;
    }

    fn turn_waypoint_right(&mut self) {
        let a = self.wx;
        let b = self.wy;
        self.wx = -b;
        self.wy = a;
    }

    fn step_waypoint(&mut self, a: &Instr) {
        let Instr(act, v) = a;
        match act {
            Action::N => self.wy -= v,
            Action::S => self.wy += v,
            Action::E => self.wx += v,
            Action::W => self.wx -= v,
            Action::L => {
                for _ in 0..(v / 90) {
                    self.turn_waypoint_left();
                }
            }
            Action::R => {
                for _ in 0..(v / 90) {
                    self.turn_waypoint_right();
                }
            }
            Action::F => {
                self.x += self.wx * v;
                self.y += self.wy * v;
            }
        }
    }
}

fn main() -> Result<(), BErr> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let re = Regex::new(r"^([A-Z])([0-9]+)$")?;

    let mut s = Ship::new();
    let mut s2 = Ship::new();

    for line in reader.lines() {
        let line = line?;
        let m = re.captures(&line).ok_or("Invalid line format")?;
        let act = match m.get(1).ok_or("Invalid line format")?.as_str() {
            "N" => Action::N,
            "S" => Action::S,
            "E" => Action::E,
            "W" => Action::W,
            "L" => Action::L,
            "R" => Action::R,
            "F" => Action::F,
            _ => return Err("Invalid action".into()),
        };
        let num = m
            .get(2)
            .ok_or("Invalid line format")?
            .as_str()
            .parse::<i32>()?;
        let a = Instr(act, num);
        s.step(&a);
        s2.step_waypoint(&a);
    }

    println!("Part 1: {}", s.x.abs() + s.y.abs());
    println!("Part 2: {}", s2.x.abs() + s2.y.abs());

    Ok(())
}
